### PR TITLE
Add support for payload-based routing of commands and requests.

### PR DESCRIPTION
**Pull request by uglybugger**
In line with our push towards [API vNext](https://github.com/OctopusDeploy/Architecture/blob/0a5980e0918906425fc20087d8704b064e9a7053/OctopusServer/Api/ApiPhilosophy.md), the Octopus API client needs to be able to cope with being given either a command or request payload and figuring out 1) where to HTTP verb it to, and 2) which HTTP verb to actually use.

This PR introduces payload-based routing, enabling things like this:

```
var command = new CreateTeamCommandV1
{
    Name = Some.ShortRandomName(),
    Description = Some.String()
};

var response = await client.Do(command, CancellationToken);  // This is the shiny part.

response.Team.SpaceId.Should().BeNull();
response.Team.Name.Should().Be(command.Name);
response.Team.Description.Should().Be(command.Description);
```

More information about HTTP routing can be found at the [HTTP subsection of the Transport-specific concerns section of the API vNext guidelines](https://github.com/OctopusDeploy/Architecture/blob/0a5980e0918906425fc20087d8704b064e9a7053/OctopusServer/Api/ApiPhilosophy.md#http).

### CODE REVIEW

File: source/Octopus.Client.Tests/PublicSurfaceAreaFixture.ThePublicSurfaceAreaShouldNotRegress..NETCore.approved.txt [hunk 1]
```diff
@@ -1488,8 +1497,10 @@ Octopus.Client.Extensions
 {
   abstract class StringExtensions
   {
-    static String CommaSeperate(IEnumerable<Object>)
-    static String NewLineSeperate(IEnumerable<Object>)
```
Comment (uglybugger) [inline source/Octopus.Client.Tests/PublicSurfaceAreaFixture.ThePublicSurfaceAreaShouldNotRegress..NETCore.approved.txt:1492]: Twitch.

File: source/Octopus.Server.Client/AppDomainScanner.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+﻿using System;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+namespace Octopus.Client
+{
+    internal static class AppDomainScanner
```
Comment (uglybugger) [inline source/Octopus.Server.Client/AppDomainScanner.cs:8]: Primitive scanner to allow the client's default HTTP route locator service implementation to find all the types implementing HTTP routes for command/request payloads.

File: source/Octopus.Server.Client/OctopusAsyncClient.cs [hunk 3]
```diff
@@ -123,12 +128,42 @@ public IOctopusSystemAsyncRepository ForSystem()
...
+            return response;
+        }
+
+        private async Task<TResponse> Send<TResponse>(HttpMethod method, Uri relativeRoute, object payload, CancellationToken cancellationToken)
+        {
+            var uri = serverEndpoint.OctopusServer.Resolve(relativeRoute.ToString());
+            var octopusRequest = new OctopusRequest(method.Method.ToUpperInvariant(), uri, payload);
+            var response = await DispatchRequest<TResponse>(octopusRequest, true, cancellationToken);
+            return response.ResponseResource;
+        }
```
Comment (uglybugger) [inline source/Octopus.Server.Client/OctopusAsyncClient.cs:159]: These methods are the meat of this change. They allow things like:
```
var response = await client.Do(new DoAThingCommand(...), ct);
```
and
```
var response = await client.Request(new AskAQuestionRequest(...), ct);
```

File: source/Octopus.Server.Client/HttpRouting/HttpRouteExtractor.cs [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            // If we're routing a GET request, every property not added to the route gets added as a query string
+            // parameter - otherwise it can just go as a JSON payload.
+            if (httpMethod == HttpMethod.Get)
+            {
+                var queryString = payloadProperties
+                    .OrderBy(kvp => kvp.Key)
+                    .Select(kvp => $"{kvp.Key.ToCamelCase()}={WebUtility.UrlEncode(kvp.Value.ToString())}")
+                    .StringJoin("&");
+
+                if (!string.IsNullOrEmpty(queryString)) route += $"?{queryString}";
```
Comment (skermajo) [inline source/Octopus.Server.Client/HttpRouting/HttpRouteExtractor.cs:146]: Shame about having to support net452.

**Review by skermajo — APPROVED** [review]
LGTM 👍 This is really great - nice interface and great developer experience. I don't care how you call the server, **I just want to do 'this'**! 🏆 

Minor nitpick would be various compiler hints in some of the new code where it could be 'modernised' (`new()`, unused using etc).  Otherwise GG 👍

File: source/Octopus.Server.Client/HttpRouting/HttpRouteExtractor.cs (hunk 4 above)
Comment (uglybugger) [reply source/Octopus.Server.Client/HttpRouting/HttpRouteExtractor.cs:146]: Ooh, yes 😛

**Pull request merged**